### PR TITLE
feat: invalidate cache

### PR DESCRIPTION
**Pull request by evilebottnawi**
Steps:
- [x] Add ability to determinate own cache identifier.
- [x] Invalidate cache after `cache-loader` updated.
- [x] Invalidate cache after `cache-loader` options changed.
- [x] Invalidate cache if environment changed.
- [x] Invalidate cache if loader in chain updated.
As sokra says:
```
"postinstall": "rm -rf .cache-loader"
```
We can just add this in README.
- [x] Search solution for loaders using fs based configuration files, maybe just add docs.
As sokra says:
```
configuration files -> loaders should add these files to their dependencies (with `this.addDependency`). dependencies' mtimes are stored by the cache-loader
```
Partial fixed: https://github.com/webpack-contrib/cache-loader/issues/3

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -58,13 +59,25 @@ function loader(...args) {
 }
 
 function pitch(remainingRequest, prevRequest, dataInput) {
-  const options = loaderUtils.getOptions(this) || {};
-  const cacheDirectory = options.cacheDirectory || path.resolve('.cache-loader');
+  const loaderOptions = loaderUtils.getOptions(this) || {};
+  const defaultOptions = {
```
Comment (STRML) [inline src/index.js]: This is going to be the same for every call to `pitch()` so it would make sense to define it outside the fn.

**Comment by michael-ciniawsky** [comment]
What's missing here ? [`babel-loader` #60](https://github.com/babel/babel-loader/pull/460) was merge a while ago 😛

File: src/index.js (hunk 1 above)
Comment (evilebottnawi) [inline src/index.js]: @STRML done

Comment (STRML) [inline src/index.js]: Thanks - doesn't appear pushed though

Comment (michael-ciniawsky) [inline src/index.js]: @evilebottnawi Can't the whole `defaultOptions` block be moved for some reason ? 🙃

Comment (evilebottnawi) [inline src/index.js]: @michael-ciniawsky oh, yep, your are right

**Comment by michael-ciniawsky** [comment]
Removing the WIP seems to be good to go ?

File: src/index.js [hunk 2]
```diff
@@ -58,13 +61,25 @@ function loader(...args) {
...
+      env: process.env.NODE_ENV || 'development',
+    }),
+  };
+  const options = Object.assign({}, defaultOptions, loaderOptions);
+  const { cacheIdentifier, cacheDirectory } = options;
   const data = dataInput;
-  data.remainingRequest = remainingRequest;
   const callback = this.async();
-  const hash = digest(remainingRequest);
+  const hash = digest(JSON.stringify({
```
Comment (sokra) [inline src/index.js]: I would change this to:

``` js
const hash = digest(remainingRequest + cacheIdentifier);
```

Faster. Options don't need to be included.

**Comment by sokra** [comment]
Thanks

**Pull request merged**